### PR TITLE
Add checks to unique_id

### PR DESCRIPTION
**Pull request by andralex**
Also make it simpler (the original template signature had additional unused template parameters).

### CODE REVIEW

File: cudax/include/cuda/experimental/__stf/utility/unique_id.cuh [hunk 1]
```diff
@@ -33,34 +33,47 @@ namespace reserved
...
-  constexpr unique_id(const int val) noexcept
+  {
+    assert(_value >= 0 && "Reading a dead unique_id is not allowed.");
+  }
+  unique_id(const int val) noexcept
       : _value(val)
-  {}
-  constexpr operator int() const noexcept
   {
+    assert(val >= 0 && "A unique_id must contain a non-negative value.");
```
Comment (miscco) [inline cudax/include/cuda/experimental/__stf/utility/unique_id.cuh:49]: Is there a reason you are removing the `constexpr from that functions?`

Comment (andralex) [reply cudax/include/cuda/experimental/__stf/utility/unique_id.cuh:49]: I think `constexpr` is superfluous in this class, but I'll add it back for the sake of "constexpr all the things".

**Review by miscco — APPROVED** [review]

**Pull request merged**